frontend/src/stores/ethereum.ts

```diff
@@ -91,57 +94,64 @@ export const useEthereumStore = defineStore('ethereum', () => {
-      if (!net) return;
-      const isSapphire = sapphire.NETWORKS[net as number];
-      signer.value = isSapphire ? sapphire.wrap(s) : s;
-      unwrappedSigner.value = s;
-      provider.value = isSapphire ? markRaw(sapphire.wrap(s.provider)) : s.provider;
-      unwrappedProvider.value = s.provider as JsonRpcProvider;
-      network.value = net;
-      address.value = addr;
-    };
-
-    const [addr, net] = await Promise.all([
-      s.getAddress(),
-      s.provider.getNetwork().then(net => networkFromChainId(net.chainId)),
-    ]);
-    setSigner(addr, net);
-
-    if (!isMetaMask) {
-      status.value = ConnectionStatus.Connected;
-      return;
+  async function init(addr: string) {
+    const eth = window.ethereum!;
+
+    const browserProvider = await new BrowserProvider(eth);
+    const providerNetwork = await browserProvider.getNetwork();
+    const chainId = networkByChainId(providerNetwork.chainId);
+
+    const isSapphire = sapphire.NETWORKS[chainId];
+
+    let sapphireProvider = null;
+
+    if (isSapphire) {
+      sapphireProvider = sapphire.wrap(browserProvider);
     }
-    ethProvider.on('accountsChanged', (accounts) => {
-      setSigner(accounts[0], network.value);
-    });
-    ethProvider.on('chainChanged', (chainId) => {
-      // setSigner(address.value, networkFromChainId(chainId));
 
-      // TODO: Dirty fix, reload the app to ensure the state resets after chain switch
+    signer.value = isSapphire
+      ? await sapphireProvider!.getSigner(addr)
+      : await browserProvider.getSigner(addr);
+
+    unwrappedSigner.value = await browserProvider.getSigner(addr);
+    provider.value = isSapphire ? markRaw(sapphireProvider!) : browserProvider;
+    unwrappedProvider.value = browserProvider;
+    network.value = chainId;
+    address.value = addr;
+  }
+
+  const connect = async () => {
+    const eth = await getEthereumProvider();
+
+    const accounts: string[] = await (eth.request?.({
+      method: 'eth_accounts',
+    }) || Promise.resolve([]));
+
+    if (!accounts || accounts?.length <= 0) {
+      throw new Error('[useEthereumStore] Request account failed!');
+    }
+
+    await init(accounts[0]);
+
+    eth.on('accountsChanged', (accountsChanged) => {
+      init(accountsChanged[0]);
+    });
+    eth.on('chainChanged', () => {
       window.location.reload();
     });
-    ethProvider.on('connect', () => (status.value = ConnectionStatus.Connected));
-    ethProvider.on('disconnect', () => (status.value = ConnectionStatus.Disconnected));
-  }
+    eth.on('connect', () => (status.value = ConnectionStatus.Connected));
+    eth.on('disconnect', () => (status.value = ConnectionStatus.Disconnected));
+  };
 
-  function checkIsCorrectNetwork() {
-    return window.ethereum.networkVersion.toString() === Network.FromConfig.toString();
+  async function checkIsCorrectNetwork() {
+    return network.value.toString() === Network.FromConfig.toString();
   }
```

Comment:
Need to revert this part of code. It should be checking the provider network.

Reply:
Actually it's fine, as:
`await eth.connect();`
Is called before, so it should set the current network value.

Reply:
As `window.ethereum.networkVersion` is deprecated.